feat(replacers) Move replacer query logic to ReplacingMergeTree step 1

fpacifici opened the pull request:
While trying to make replacements work on Errors it seemed that the logic building the INSERT query in replacement is not in the right place. A lot of this logic is really specific to ReplacingMergeTrees, thus it should go where the ReplacingMergeTree is. 
Specifically:
- the fact that deletions and updates are done through INSERT
- the version column/ deleted column
- the required columns for the deletions.

This is just the first step in that direction and removes the required columns from the dataset level. More to come, but in the meantime this unblocks replacements on errors as well since I am not sure how long the rest of the refactoring will be.

fpacifici committed:
Move required columns into replacing merge tree

tkaemming commented on snuba/datasets/events.py:
```diff
@@ -260,6 +260,14 @@ def __init__(self) -> None:
+                "timestamp",
+                "deleted",
+                "retention_days",
+            ],
```
I think it's important to be able to clearly articulate why columns end up in this set. Ideally this set wouldn't even need to be explicitly defined like this in the future, since I think most of these are derived from other aspects of the schema definition.

My interpretation:

- `project_id` and `event_id` are essentially a composite primary key, so both fields need to be in here, otherwise the tombstoning strategy wouldn't work. These columns are both in the order by clause, which is the closest thing we have to a primary key, along with `timestamp`.
- `timestamp` and `retention_days` are in the partition key, so they need to remain, otherwise the tombstone record would end up in a different partition and never be merged.

It seems to me that it follows that these fields also need to be treated as generally immutable.

`deleted` and `group_id` don't have these same constraints, and I assume these are in here because these fields are not nullable in the schema definition.

- `deleted` is a special case, since [the value isn't carried over from the original row in deletions](https://github.com/getsentry/snuba/blob/master/snuba/replacer.py#L197). Should it be in here at all? It seems like it'd be more effective for this column to be defined separately.
- Should `group_id` be just be written to a static value like `0` that isn't part of the valid input domain? It doesn't seem necessary to keep the original value. I don't think we need to this now, but maybe worth considering. In general I think we should try and avoid preserving as many fields from the original row when deleting as possible.

fpacifici replied:
I picked these fields to preserve compatibility with the current deletion process. Essentially the deletion now is:
```
INSERT INTO sentry_dist (event_id, project_id, group_id, timestamp, deleted, retention_days)
SELECT event_id, project_id, group_id, timestamp, 1, retention_days
WHERE ....
```

I don't really know why group_id is there with the original value, but this is how replacer works now. https://github.com/getsentry/snuba/blob/master/snuba/replacer.py#L197

I am ok splitting required columns from primary key columns, but I am not sure about changing the behavior of group_id. 
About delete, I will actually remove it from here and let the replacer figure it out. Anyway it will be taken care of when I introduce prepared statements in step 3

tkaemming commented on snuba/datasets/errors.py:
```diff
@@ -155,6 +155,14 @@ def __init__(self) -> None:
+                "timestamp",
+                "deleted",
+                "retention_days",
+            ],
```
See comment below for more context, but by following the same logic from the definition in `EventsDataset`, it's unclear to me where this column set comes from considering this table layout is considerably different and this table contains far less nullable fields.

fpacifici replied:
Indeed there are more required columns to add.
Unfortunately I cannot test the replacement on that table till https://github.com/getsentry/snuba/pull/774 so I could not see that.

tkaemming reviewed (changes requested):
I like the idea here, but don't understand how this actually works for errors.

fpacifici committed:
Fetch required columns from the schema

fpacifici committed:
Add comment

fpacifici commented:
Removed the need to explicitly list the required columns to use for deletion. Now we pick them up from the schema.
Which can break if someone makes manual changes to the clickhouse table, but that was the case before as well, and actually it seems fair

fpacifici commented on snuba/clickhouse/columns.py:
```diff
@@ -68,10 +69,31 @@ def flatten(self, name: str) -> Sequence[FlattenedColumn]:
 
 
-class Nullable(ColumnType):
+class ColumnTypeWithModifier(ABC, ColumnType):
```
I think a modifier should become an attribute of a column, and not a type in its own, but that's an orthogonal refactoring.

fpacifici commented on snuba/datasets/events.py:
```diff
@@ -198,7 +199,7 @@ def __init__(self) -> None:
                 # other tags
                 ("tags", Nested([("key", String()), ("value", String())])),
-                ("_tags_flattened", String()),
+                ("_tags_flattened", WithDefault(String(), "''")),
```
This was wrong.

tkaemming commented on snuba/datasets/schemas/tables.py:
```diff
@@ -255,6 +261,33 @@ def __init__(
+        This delas with flattened columns (since the default iteration behavior on
+        ColumnSet is to return flattened columns, thus this approach seems more robust).
+        This means there are no Nested columns, only Arrays.
+        """
```
I don't think this is necessarily accurate, [looking at the documentation](https://clickhouse.tech/docs/en/query_language/insert_into/):

> The query can specify a list of columns to insert [(c1, c2, c3)]. In this case, the rest of the columns are filled with:
> - The values calculated from the DEFAULT expressions specified in the table definition.
> - Zeros and empty strings, if DEFAULT expressions are not defined.

This doesn't say anything about arrays (looking at the comment below) but I assume if there is no default expression, they just become empty. It doesn't actually say anything about `Nullable` columns either, but I assume those would become `NULL` instead of the non-`NULL` default.

With this in mind (assuming we're not running with `strict_insert_defaults` enabled), I don't think whether or not a column is `Nullable` has a significant impact on whether or not it should be included in columns that are preserved in the deletion query. In fact, if we left out `group_id` from the columns to preserve during deletion, it would have actually ended up doing the thing that I mentioned [in this comment](https://github.com/getsentry/snuba/pull/778#discussion_r379600079), which makes me even more confused about how it ended up in there in the first place.[1]

I think that the only columns that need to be preserved here during are the columns that affect the viability of a row to be merged during optimization — specifically rows that are referenced in the `ORDER BY` and `PARTITION BY` DDL clauses.

In retrospect, I realize the "this table contains far less nullable fields" part of [this comment](https://github.com/getsentry/snuba/pull/778#discussion_r379605521) was a bit misleading because I misunderstood how non-`Nullable` columns are treated with `INSERT INTO` and expected it to behave more like PostgreSQL or MySQL.

1: One weird side effect of these replacements is that they will trash the min side of a min/max index since they are not conditional on `deleted = 1`.

fpacifici replied:
damn, that's why we did not receive complaints of replacements not working after introducing flattened_tags. The column is not nullable, we were not providing anything so it was adding the default .....

```
I think that the only columns that need to be preserved here during are the columns that affect the viability of a row to be merged during optimization — specifically columns that are referenced in the ORDER BY and PARTITION BY DDL clauses.
```

Those plus the deleted column. So I guess I will go back to the original implementation with the only difference of splitting order by columns and partition by columns.

fpacifici committed:
Revert changes from previous commit

fpacifici committed:
Return the list of mandatory columns

fpacifici commented:
Reverted to explicitly listing the columns for tombstones in the schema, though they are now split between partition columns and order by columns and version column.

fpacifici committed:
Merge branch 'master' into feat/replacer_into_merge_tree

fpacifici commented:
I think https://github.com/getsentry/snuba/pull/781 is a much better approach to this problem.
We can reconsider building a fully generic replacer concept based on parameters we store in ReplacingMergeTree but I believe it would be an overengineered solution as of now where we only have really one use case for replacements: Events (errors are really the same use cases except for some parameters).
